{% extends 'base.html' %}

{% block header %}

<style>
	.card:hover {
		box-shadow: 0 8px 16px 0 rgba(0,0,0,0.2);
		transform: scale(1.005);
	}
</style>

{% endblock %}

{% block content %}
	<div class="row text-center mx-auto">
		<div class="col-md-1">

		</div>
		<div class="col-md-10 pt-4">
			<div class="p-4">
				Welcome to the webserver of NTM-Profiler - a pipeline which allows users
				to analyse <i>Mycobacterial</i> whole genome sequencing data to
				predict lineage and drug resistance. Follow the instructions below to
				upload a new sample or view analysed runs.
			</div>



			<h5 class="">How does it work?</h5>
			<div class="pb-4">
				The pipeline searches for small variants and big deletions associated
				with drug resistance. It will also report the lineage. By default it
				uses Trimmomatic to trim the reads, BWA (or minimap2 for nanopore) to
				align to the reference genome and GATK (open source v4) to call variants.
			</div>


			<div class="row justify-content-md-center">
				<div class="col-md-6 pb-4">
					<div class="card border-dark text-center">
					  <div class="card-header bg-dark text-white">
					    Step 1
					  </div>
					  <div class="card-body">
					    <h5 class="card-title">Profile your sample</h5>

							<p class="card-text">
								Upload your next generation sequencing data in <b>fastQ</b> format.
								You can upload one or two (forward and reverse) fastq files. When you
								upload your data, the run will be be assigned a unique ID.
								Please take a note of this ID as you will need to to find your results later.
								Batch upload of samples is also possible.
							</p>


					    <a class="btn btn-dark" href="{{ url_for('main.upload') }}">Upload</a>
					  </div>
					</div>
				</div>
			</div>


			<div class="row justify-content-md-center">
				<div class="col-md-6 pb-4">
					<div class="card border-dark text-center">
					  <div class="card-header bg-dark text-white">
					    Step 2
					  </div>
					  <div class="card-body">
					    <h5 class="card-title">View the results</h5>

									<p class="card-text">
										Find your results by entering you unique run ID directly into the search box below.
									</p>

							<form method="post" enctype=multipart/form-data>
								<div class="row justify-content-md-center">
									<div class="col-md-6">
										<input type="text" name="sample_id" class="form-control" placeholder="Sample ID">
									</div>
								</div>
								<div class="row justify-content-md-center pt-3">
									<div class="col-md-6">
										<button type="submit" class="btn btn-dark">Submit</button>
									</div>
								</div>
							</form>
					  </div>
					</div>
				</div>
			</div>



			<h5 class="pt-4">Which mutations are being used to prodict resistance?</h5>
			<div class="pb-4">
				Resistance markers have been retreived from the literature.
				You can take a look at the
				up to date CSV files <a href="https://github.com/ntm-db/ntm-db/tree/master/db">here</a> by navigating to the
				relevant species folder and viewing the <code>variants.csv</code> file.
			</div>



			<h5 class="pt-4">Can I run NTM-Profiler on my own computer?</h5>
			<div class="pb-4">
				If you have access to a linux or macOS operating system then you can download
				the commandline version of NTM-Profiler. For more information on this please
				visit the	<a href="https://github.com/jodyphelan/ntm-profiler/">github repository</a>.
				If you would like to create your own instance of this webserver, the code
				is fully open-source and available at this <a href="https://github.com/jodyphelan/ntm-profiler-webserver/">github repository</a>
			</div>

			<h5 class="pt-4">Give feedback or request a feature</h5>
			<div class="pb-4">
				We are always looking to improve NTM-Profiler. If you see some results that
				don't look right or would like to see some additional functionality in the
				future, don't hesitate to get in touch via the
				<a href="https://github.com/jodyphelan/ntm-profiler-webserver/issues"></a>
				section on our github page.
			</div>

			<h5 class="pt-4">Citation</h5>
			<div class="pb-4">
				Please cite this website if you use it in your research.
			</div>


			<hr>
			<h6 class="">Disclaimer</h6>
			<div class="">
				This tool is for Research Use Only. It has not been approved, cleared,
				or licensed by any regulatory authority. By submitting sequence data the
				 user acknowledges no intended medical purpose or objective such as
				 clinical diagnosis, patient management, or human clinical trials.
			</div>
		</div>
	</div>

{% endblock %}
